pyrlang/node.py

```diff
@@ -369,64 +376,135 @@ def link_nowait(self, pid1, pid2, local_only=False):
-            link them. Assume remote process handles its own linking.
+    async def link(self, from_pid, to_pid, local_only=False):
+        """ Process a link event generated by from_pid, targeting to_pid.
 
-            :param pid1: First pid
-            :type pid1: term.pid.Pid
-            :param pid2: Second pid
-            :type pid2: term.pid.Pid
-            :param local_only: If set to True, linking to remote pids will send
-                LINK message over dist_proto protocol
+            :param from_pid: Source pid
+            :type from_pid: term.pid.Pid
+            :param to_pid: Target pid
+            :type to_pid: term.pid.Pid
+            :param local_only: Deprecated, ignored.
         """
-        if pid1.is_local_to(self):
-            if pid1 in self.processes_:
-                self.processes_[pid1].add_link(pid2)
-            else:
-                # not exists
-                await self._send_exit_signal(pid2, pid1, Atom("noproc"))
+        if from_pid.is_local_to(self) and from_pid not in self.processes_:
+            LOG.info("local from_pid %s not found during link", from_pid)
+            return
+        if to_pid.is_local_to(self) and to_pid not in self.processes_:
+            await self._send_exit_signal(to_pid, from_pid, Atom("noproc"))
+            return
 
-        elif not local_only:
-            link_m = ('link', pid2, pid1)
-            await self.dist_command(receiver_node=pid1.node_name_,
-                                    message=link_m)
+        from_process = self.processes_.get(from_pid, None)
+        to_process = self.processes_.get(to_pid, None)
 
-        if pid2.is_local_to(self):
-            if pid2 in self.processes_:
-                self.processes_[pid2].add_link(pid1)
-            else:
-                # not exists
-                await self._send_exit_signal(pid1, pid2, Atom("noproc"))
+        if from_pid.is_local_to(self) and to_pid.is_local_to(self):
+            from_process._links[to_pid] = True
+            to_process._links[from_pid] = True
 
-        elif not local_only:
-            link_m = ('link', pid1, pid2)
-            await self.dist_command(receiver_node=pid2.node_name_,
-                                    message=link_m)
+        elif from_pid.is_local_to(self):  # we're generating the message
+            link_exists = from_process._links.get(to_pid, False)
+            if link_exists is True:
+                return
 
-    async def unlink(self, pid1, pid2, local_only=False):
+            from_process._links[to_pid] = True
+            await self.dist_command(receiver_node=to_pid.node_name_,
+                                    message=('link', from_pid, to_pid))
+
+        elif to_pid.is_local_to(self):
+            if from_pid in to_process._links:
+                return
+            to_process._links[from_pid] = True
+
+        else:
+            raise ValueError("Illegal to generate remote-remote links")
```

Comment:
Same as the other one, we should just look that we get nice cleanup if these occur

Reply:
I'll change this to a warning, in keeping with what ERTS does: https://github.com/erlang/otp/blob/master/erts/emulator/beam/dist.c#L3012C2-L3012C18 .